File: Sources/APTradeDomain/Trade.swift
import Foundation

public enum TradeSide: String, Codable, Sendable {
    case buy, sell, dividend
}

public enum TradeError: Error, Equatable, Sendable {
    case insufficientFunds
    case insufficientShares
    case invalidQuantity
}

public struct Transaction: Identifiable, Equatable, Codable, Sendable {
    public let id: UUID
    public let symbol: String
    public let side: TradeSide
    public let quantity: Quantity
    public let price: Money
    public let date: Date
    public let pieId: String?
    public let isDrip: Bool

    public init(id: UUID = UUID(), symbol: String, side: TradeSide,
                quantity: Quantity, price: Money, date: Date, pieId: String? = nil,
                isDrip: Bool = false) {
        self.id = id
        self.symbol = symbol
        self.side = side
        self.quantity = quantity
        self.price = price
        self.date = date
        self.pieId = pieId
        self.isDrip = isDrip
    }

    /// Custom decoding so ledgers persisted before M8 (no `pieId`/`isDrip` keys)
    /// decode unchanged, defaulting `pieId` to `nil` and `isDrip` to `false`.
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(UUID.self, forKey: .id)
        symbol = try container.decode(String.self, forKey: .symbol)
        side = try container.decode(TradeSide.self, forKey: .side)
        quantity = try container.decode(Quantity.self, forKey: .quantity)
        price = try container.decode(Money.self, forKey: .price)
        date = try container.decode(Date.self, forKey: .date)
        pieId = try container.decodeIfPresent(String.self, forKey: .pieId)
        isDrip = try container.decodeIfPresent(Bool.self, forKey: .isDrip) ?? false
    }
}

File: shared/src/commonMain/kotlin/com/aptrade/shared/domain/Trade.kt
package com.aptrade.shared.domain

import com.ionspin.kotlin.bignum.decimal.BigDecimal
import kotlin.random.Random

enum class TradeSide { Buy, Sell, Dividend }

sealed class TradeError(message: String) : Exception(message) {
    object InsufficientFunds : TradeError("Insufficient funds")
    object InsufficientShares : TradeError("Insufficient shares")
    object InvalidQuantity : TradeError("Invalid quantity")
}

/** One executed paper trade. `id` is caller-supplied (deterministic in tests).
 *  `pieId` is an optional attribution tag: non-null when the trade was executed as part
 *  of an Investment Plan (pie) contribution or rebalance, null for ordinary manual trades.
 *  Pure metadata — never an input to cash/position/realized-P&L math. Transcribed from
 *  Sources/APTradeDomain/Trade.swift. */
data class Transaction(
    val id: String,
    val symbol: String,
    val side: TradeSide,
    val quantity: BigDecimal,
    val price: Money,
    val epochSeconds: Long,
    val pieId: String? = null,
    val isDrip: Boolean = false,
)

/** Unique-enough trade id without a platform UUID or kotlinx.datetime dependency.
 *  128 bits of randomness — collision-negligible, process-stable. */
fun generateTradeId(): String =
    "txn-${Random.nextLong().toULong().toString(16)}-${Random.nextLong().toULong().toString(16)}"
